storagenode/piecestore/endpoint.go

```diff
@@ -647,17 +648,48 @@ func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainReques
 		return nil, status.Error(codes.InvalidArgument, Error.Wrap(err).Error())
 	}
 
-	numDeleted := 0
-
 	// subtract some time to leave room for clock difference between the satellite and storage node
 	createdBefore := retainReq.GetCreationDate().Add(-endpoint.config.RetainTimeBuffer)
 
+	go func(satelliteID storj.NodeID, createdBefore time.Time, filter *bloomfilter.Filter) {
+		if endpoint.retainCtx != nil {
+			select {
+			case <-endpoint.retainCtx.Done():
+				endpoint.log.Error("retain error", zap.Error(endpoint.retainCtx.Err()))
+				return
+			default:
+			}
+
+			err := endpoint.RetainPieces(endpoint.retainCtx, satelliteID, createdBefore, filter)
+			if err != nil {
+				endpoint.log.Error("retain error", zap.Error(err))
+			}
+		} else {
+			endpoint.log.Error("No retain context. Stopping garbage collection.")
+		}
+
+	}(peer.ID, createdBefore, filter)
+
+	return &pb.RetainResponse{}, nil
+}
+
+// RetainPieces is called in a goroutine by Retain
+func (endpoint *Endpoint) RetainPieces(ctx context.Context, satelliteID storj.NodeID, createdBefore time.Time, filter *bloomfilter.Filter) (err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
It could be worth it to add some parameters to `mon.Task()`

E.g `mon.Task()(&ctx, satelliteID, ceatedBefore, filter.Size())(&err)`